gelf/tcpreader.go

```diff
@@ -13,43 +14,90 @@ type TCPReader struct {
 	messages chan []byte
 }
 
-func newTCPReader(addr string) (*TCPReader, chan string, error) {
+type conn_channels struct {
+	drop    chan string
+	confirm chan string
+}
+
+func newTCPReader(addr string) (*TCPReader, chan string, chan string, error) {
 	var err error
 	tcpAddr, err := net.ResolveTCPAddr("tcp", addr)
 	if err != nil {
-		return nil, nil, fmt.Errorf("ResolveTCPAddr('%s'): %s", addr, err)
+		return nil, nil, nil, fmt.Errorf("ResolveTCPAddr('%s'): %s", addr, err)
 	}
 
 	listener, err := net.ListenTCP("tcp", tcpAddr)
 	if err != nil {
-		return nil, nil, fmt.Errorf("ListenTCP: %s", err)
+		return nil, nil, nil, fmt.Errorf("ListenTCP: %s", err)
 	}
 
 	r := &TCPReader{
 		listener: listener,
 		messages: make(chan []byte, 100), // Make a buffered channel with at most 100 messages
 	}
 
-	signal := make(chan string, 1)
+	close_signal := make(chan string, 1)
+	done_signal := make(chan string, 1)
 
-	go r.listenUntilCloseSignal(signal)
+	go r.listenUntilCloseSignal(close_signal, done_signal)
 
-	return r, signal, nil
+	return r, close_signal, done_signal, nil
 }
 
-func (r *TCPReader) listenUntilCloseSignal(signal chan string) {
-	defer func() { signal <- "done" }()
-	defer r.listener.Close()
+func (r *TCPReader) accepter(connections chan net.Conn) {
 	for {
 		conn, err := r.listener.Accept()
 		if err != nil {
 			break
```

Comment:
If an error occurs here, I think this will have no effect on the program but to silently stop the acceptor.

Reply:
This is the case, the thing with the whole `tcpreader` is to be a test harness, that is the reason all the functions are unexported. I should probably rename it so that it is clear that it is a test harness. It could probably be refactored and cleaned up to expose the API, but I am not aware of any downstream projects that require a GELF TCP Reader.

Reply:
Ah yes, I forgot about that.
It's fine then 👍